fix: Fix L1 publishing

PhilWindle opened the pull request:
This fixes a bug in the L1 Tx Publisher. The bug manifests as:

1. A tx is sent to L1. The publisher monitors it for a period of time (until it would no longer be valid).
2. The publisher believes it has timed out, but it hasn't. There is a race condition and it has actually been mined.
3. The publisher submits and monitors cancellation transactions.
4. It no longer monitors the original transaction and so remains stuck in a terminal state of 'cancelling'.

spalladino commented on yarn-project/ethereum/src/l1_tx_utils.test.ts:
```diff
@@ -579,6 +592,94 @@ describe('L1TxUtils', () => {
+      // We need to intercept the call to send a transaction to L1.
+      // We let the first one through but no more.
+      // This blocks any cancellations
+      using _1 = jest
+        .spyOn(l1Client, 'sendRawTransaction')
+        .mockImplementationOnce(async arg => {
+          // This is the actual transaction
+          const sentTx = { ...txBeingSigned! };
+          const hash = await originalSendRawTransaction.call(this, arg);
+          sentTxs.set(hash, sentTx);
+          return hash;
+        })
+        .mockImplementation(_arg => {
+          // Do nothing, there are any/all cancellations
+          const sentTx = txBeingSigned!;
+          const hash = randomBytes(32).toString('hex') as Hex;
+          sentTxs.set(hash, sentTx);
+          cancellationSent = true;
+          return Promise.resolve(hash);
+        });
```
I think we should be able to just deserialize the raw tx in `arg` rather than capture it by spying on `signTransaction`. But if not, and this works, I'm fine with it.

PhilWindle replied:
I tried that originaly, but the de-serialized tx had no nonce. Couldn't figure out what was wrong so I opted for this approach.

spalladino commented on yarn-project/ethereum/src/l1_tx_utils.ts:
```diff
@@ -777,7 +779,7 @@ export class L1TxUtils extends ReadOnlyL1TxUtils {
         const currentNonce = await this.client.getTransactionCount({ address: account });
         if (currentNonce > nonce) {
-          for (const hash of txHashes) {
+          for (const hash of allVersions) {
```
Should we set state to `MINED` and throw if we looped over `allVersions` and not found the receipt? That would have fixed the issue as well, and will also catch any instances where we fail to populate `allVersions` with all possible hashes (eg node going down).

PhilWindle replied:
Yes, anything that adds resiliency to this I think is a good idea.

PhilWindle replied:
Though, doesn't this `const receipt = await this.client.getTransactionReceipt({ hash });` only return a receipt for mined transactions. If we don't find one wouldn't that just mean nothing is yet mined?

spalladino replied:
If we don't find a receipt after looping over all hashes in `allVersions`, but `currentNonce > nonce`, it means that _some_ tx from our sender with the same nonce was mined, but none that we were tracking. And none of the ones we were tracking will get mined, since another one with that nonce made it.

FWIW some libraries (`viem` included when you `waitForTxReceipt`) will actually download all mined tx hashes for every single block and check sender/nonce, to see if your tx was replaced and by what. This is expensive in terms of rpc usage but required in UIs because wallets do not report speed-up txs back to the app, so the app has [no other way to detect the replacement.](https://github.com/ethereum/execution-apis/issues/494) But in scenarios where we control every tx that is sent like this one, we shouldn't need this, barring some bug in our code.

PhilWindle replied:
Ah yes, I missed the nonce check!

alexghr reviewed (approved):
The code looks good! The only things that's not clear to me is the  node's/viem's `getTransaction` real behaviour in the wild. Will it work with all txs for the same nonce or throw for some of them?

PhilWindle committed:
fix: Fix L1 publishing

This fixes a bug in the L1 Tx Publisher. The bug manifests as:

1. A tx is sent to L1. The publisher monitors it for a period of time (until it would no longer be valid).
2. The publisher believes it has timed out, but it hasn't. There is a race condition and it has actually been mined.
3. The publisher submits and monitors cancellation transactions.
4. It no longer monitors the original transaction and so remains stuck in a terminal state of 'cancelling'.
